From 9673c5ecba179782c2c2478dbfb6ca8da0a7cc8a Mon Sep 17 00:00:00 2001
From: Mark Fasheh <mfasheh@suse.com>
Date: Mon, 29 Aug 2011 13:38:50 -0700
Subject: [PATCH 10/20] btrfs: go readonly on insert error in
 btrfs_add_root_ref()
Patch-mainline: 3.2?

In btrfs_add_root_ref() we BUG if an error is encountered during
REF/BACKREF insertion. This does not look like a logic error, thus the BUG
is not called for. However, I don't think there's a simple way to recover
from such an error at that point, so we mark the fs readonly instead.

At the same time, we can update the following caller of
btrfs_add_root_ref() which BUG_ON from an error:

- create_subvol: now passes the return code back to userspace

- create_pending_snapshot: goes readonly on any error from
  btrfs_add_root_ref().

Signed-off-by: Mark Fasheh <mfasheh@suse.de>
---
 fs/btrfs/ioctl.c       |    4 ++--
 fs/btrfs/root-tree.c   |    8 ++++++--
 fs/btrfs/transaction.c |    2 +-
 3 files changed, 9 insertions(+), 5 deletions(-)

Index: linux-3.0-SLE11-SP2-3.0/fs/btrfs/ioctl.c
===================================================================
--- linux-3.0-SLE11-SP2-3.0.orig/fs/btrfs/ioctl.c
+++ linux-3.0-SLE11-SP2-3.0/fs/btrfs/ioctl.c
@@ -437,8 +437,8 @@ static noinline int create_subvol(struct
 	ret = btrfs_add_root_ref(trans, root->fs_info->tree_root,
 				 objectid, root->root_key.objectid,
 				 btrfs_ino(dir), index, name, namelen);
-
-	BUG_ON(ret);
+	if (ret)
+		goto fail;
 
 	inode = btrfs_lookup_dentry(dir, dentry);
 	if (IS_ERR(inode)) {
Index: linux-3.0-SLE11-SP2-3.0/fs/btrfs/root-tree.c
===================================================================
--- linux-3.0-SLE11-SP2-3.0.orig/fs/btrfs/root-tree.c
+++ linux-3.0-SLE11-SP2-3.0/fs/btrfs/root-tree.c
@@ -407,7 +407,10 @@ int btrfs_add_root_ref(struct btrfs_tran
 again:
 	ret = btrfs_insert_empty_item(trans, tree_root, path, &key,
 				      sizeof(*ref) + name_len);
-	BUG_ON(ret);
+	if (ret) {
+		btrfs_std_error(tree_root->fs_info, ret);
+		goto out_free;
+	}
 
 	leaf = path->nodes[0];
 	ref = btrfs_item_ptr(leaf, path->slots[0], struct btrfs_root_ref);
@@ -426,8 +429,9 @@ again:
 		goto again;
 	}
 
+out_free:
 	btrfs_free_path(path);
-	return 0;
+	return ret;
 }
 
 /*
Index: linux-3.0-SLE11-SP2-3.0/fs/btrfs/transaction.c
===================================================================
--- linux-3.0-SLE11-SP2-3.0.orig/fs/btrfs/transaction.c
+++ linux-3.0-SLE11-SP2-3.0/fs/btrfs/transaction.c
@@ -996,7 +996,7 @@ static noinline int create_pending_snaps
 				 parent_root->root_key.objectid,
 				 btrfs_ino(parent_inode), index,
 				 dentry->d_name.name, dentry->d_name.len);
-	BUG_ON(ret);
+	btrfs_std_error(fs_info, ret);
 	dput(parent);
 
 	key.offset = (u64)-1;
